README.md
# clojure_stat

A Clojure library designed to summarize data.  Summarizing a dataset is one of the most fundamental operations you can do in order to detect patterns and trends. Clojure is a great, general purpose programming language that can be very useful in data analytics. In this repository, I introduce a function that allows a user to summarize a dataset using custom partitions.

## Usage

This small library was intended for ad-hoc data analytics using the Clojure programming language.

## License
Distributed under the Eclipse Public License either version 1.0 or (at
your option) any later version.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2017-05-16
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2017-05-16
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://github.com/your-name/clojure_stat/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/clojure_stat/compare/0.1.0...0.1.1


test/clojure_stat/core_test.clj
(ns clojure-stat.core-test
  (:require [clojure.test :refer :all]
            [clojure-stat.core :refer :all]))

(deftest mean-test
  (testing "mean Function"
    (is (= 10 (mean 8 9 10 11 12)) "testing multiple arguments")    
    (is (= 10 (mean 10)) "testing 1 arg")        
    (is (= nil (mean)) "testing 0 args")        
    (is (= -10 (mean -8 -9 -10 -11 -12)) "testing negative numbers")        
    (is (= 0 (mean 0)) "testing zero using 1 arg")        
    (is (= 0 (mean -1 1)) "testing zero using 2 args")        
    )
  )

(deftest median-test
  (testing "median Function"
    (is (= 7 (median 5 4 9 10 0 -5 100 200)) "Even Number of arguments - take avg of 5 and 9")
    (is (= 5 (median 5 4 9 10 0 -5 100)) "Odd Number of arguments")
    (is (= 5 (median 5)) "One arg")
    (is (= nil (median)) "No args")
    (is (= -5 (median -5 -9 -3 -2 -10)) "Negative args, even number of args")
    (is (= -7 (median -5 -9 -3 -2 -10 -100)) "Negative args, odd number of args")
    )
  )

(deftest percentile-test
  (testing "percentile function"
    (is (= 1 (percentile 10 5 4 3 2 1 10 9 8 7 6)) "test whole number 10 percentile")
    (is (= 1 (percentile 0.1 5 4 3 2 1 10 9 8 7 6)) "test decimal 0.1 percentile")
    (is (= nil (percentile 10)) "no numbers in sequence")
    (is (= 55 (percentile 10 55)) "only 1 number in sequence")
    (is (= -25 (percentile 10 -5 -4 -3 -2 -1 -25 -9 -8 -7 -6)) "test negative numbers 10 percentile")
    (is (= -9 (percentile 20 -5 -4 -3 -2 -1 -25 -9 -8 -7 -6)) "test negative numbers 10 percentile")
    )
  )

(deftest summarize-test
  (let [sample_data [
          {:dt 20110331 :name "John" :age 22 :salary 65000 :education 12}
          {:dt 20120331 :name "John" :age 23 :salary 65000 :education 12}
          {:dt 20110331 :name "Jane" :age 22 :salary 100000 :education 16}
          {:dt 20120331 :name "Jane" :age 23 :salary 101000 :education 16}
          {:dt 20110331 :name "Eric" :age 22 :salary 25000 :education 12}
          {:dt 20120331 :name "Eric" :age 23 :salary 25000 :education 13}
          ]
        ]

    (testing "2 statistics"
      (let [result (summarize sample_data [:dt :education] [[:salary mean "avg_salary"] [:salary + "aggregate_income"]])]
        (is (= (mean 65000 25000) (:avg_salary (first result))))
        (is (= (+ 65000 25000) (:aggregate_income (first result))))      
        (is (= (count result) 5) )
        )
    )

    (testing "1 statistic only. Statistic is in a nested vector."
      (let [result (summarize sample_data [:education :age] [[:salary + "aggregate_income"]])]
        (is (= (+ 65000 25000) (:aggregate_income (first result))) )
        (is (= (count result) 5) )
        )
    )

    (testing "1 statistic only. Statistic is NOT in a nested vector."
      (let [result (summarize sample_data [:education :age] [:salary + "aggregate_income"])]
        (is (= (+ 65000 25000) (:aggregate_income (first result))) )
        (is (= (count result) 5) )
        )
    )

    (testing "1 statistic only but no group-by arguments"
      (let [
        result (summarize sample_data [:salary + "aggregate_income"])
        agg_inc (reduce (fn [x y] (+ x y)) (map :salary sample_data))
        ]
        (is (= agg_inc (:aggregate_income result)) 
          "Summarize returns a simple map, not a vector with one element, which is why we don't call (first result)
           like the other tests")
        )
    )

    )
  )





src/clojure_stat/core.clj
(ns clojure-stat.core)

(defn- verify-stats-input-for-summarize [stats]
  "This is a private helper function for the summarize macro.
   If data has three elements and the first element is not a vector, then wrap the three elements in a vector,
   and nest that vector in another vector. Otherwise, return the stats vector as is.
  "
  (cond
    (and (= (count stats) 3) (not (vector? (first stats)))) (vector stats)
    :else stats 
    )
  )

(defmacro summarize
  "Takes a vector of maps (data), a vector of keys (groups), and a vector of statistics (stats).  The function applies the summary stats in (stats) to
   each partition that is created from the keys in (groups).
   Sample data and macro call proceed:
   (def sample_data [
          {:dt 20110331 :name \"John\" :age 22 :salary 65000 :education 12}
          {:dt 20120331 :name \"John\" :age 23 :salary 65000 :education 12}
          {:dt 20110331 :name \"Jane\" :age 22 :salary 100000 :education 16}
          {:dt 20120331 :name \"Jane\" :age 23 :salary 101000 :education 16}
          {:dt 20110331 :name \"Eric\" :age 22 :salary 25000 :education 12}
          {:dt 20120331 :name \"Eric\" :age 23 :salary 25000 :education 12}
          ])
   The following will summarize the data by :dt and :education. The macro will apply the mean function to :salary and 
   place the result in a key named avg_salary.  The macro will apply the + function to :salary and place the result
   in a key named aggregate_income
   (summarize sample_data [:dt :education] [[:salary mean \"avg_salary\"] [:salary + \"aggregate_income\"]])
   other examples:   
   (summarize sample_data [:education :age] [[:salary + \"aggregate_income\"]])
   (summarize sample_data [:education :age] [:salary + \"aggregate_income\"])
   (summarize sample_data [[:salary + \"aggregate_income\"]])
   (summarize sample_data [:salary + \"aggregate_income\"])
  "
  ([data group stats]
   `(let [sorted-data# (sort-by (juxt ~@group) ~data)
          partitioned-data# (partition-by #(apply vector (map (fn [attr#] (attr# %)) ~group)) sorted-data#)]
      (map
       (fn [partition#]
         (let [first_record# (first partition#)
               grouping# (reduce (fn [result# obs#] (assoc result# obs# (obs# first_record#))) {} ~group)
               stats# (clojure-stat.core/summarize partition# ~stats)]
           (merge grouping# stats#)
           )
        )
       partitioned-data#
       )
      )
   )
   ([data stats]
     (do 
       ;;verify-stats-input-for-summarize is a private function, so you should'nt put it in the s-expression,
       ;;since it will likely get macroexpanded in a different namespace.
       (let [verified-stats (verify-stats-input-for-summarize stats)]
        `(let [
              stats# (reduce (fn [result# [item# fn# name#]] (conj result# (keyword name#) (apply fn# (map item# ~data)))) [] ~verified-stats)
              ]
          (apply hash-map stats#)
          )
       )
     )
    )
  )

(defn mean [& data]
  "Takes a sequence of numbers and returns the mean. Sample uses are (mean 1 2 3 4) and (apply mean [1 2 3 4])."
  (let [cnt (count data)]
    (if (> cnt 0)
      (with-precision 10 (/ (apply + data) cnt))
      nil
      )
    )
  )

(defn median [& data]
  "Takes a sequence of numbers and returns the median. Smaple uses are (median 1 2 3 4 5) and (apply median [1 2 3 4 5])"
  (let [cnt (count data)]
    (if (> cnt 0)
      (let [sorted-data (sort data)]
        (cond
          (= cnt 1) (first data)
          :else (if (odd? cnt)
                  (nth sorted-data (/ (dec cnt) 2))
                  (mean (nth sorted-data (/ cnt 2)) (nth sorted-data (dec (/ cnt 2))))
                  ))))
    )
  )

(defn percentile [pct & data]
  (let [total-records (count data)
        sorted-data (sort data)
        adj-pct (cond (<= pct 0) 0.01 
                      (and (>= pct 1) (<= pct 100)) (/ pct 100)
                      (> pct 100) 1
                      :else pct
                )
       ]
       (cond
        (= total-records 0) nil
        (= total-records 1) (first data)
        :else (nth sorted-data (dec (int (Math/ceil (* total-records adj-pct)))))
         )
       )
  )
